gatsby/content/blog/2022/10/2022-10-18-faster-joins.mdx

```diff
@@ -0,0 +1,43 @@
+
+It sounds simple enough, but in reality this has meant a lot of work on Synapse. The main problem was that huge amounts of the Synapse codebase assumed that we either had all the state for a room, or none of it. That assumption no longer holds, which meant lots of changes to the code to handle "partial state". Along the way we've also had to make some other improvements: for example, we [found and fixed](https://github.com/matrix-org/synapse/security/advisories/GHSA-jhjh-776m-4765) some long-standing bugs in event authorisation, and implemented [cancellation](https://github.com/matrix-org/synapse/issues/3528) for requests where the client disconnects before we finish handling the request. For an idea of the things being changed at the Matrix protocol level, take a look at [MSC3902](https://github.com/matrix-org/matrix-spec-proposals/pull/3902).
+
+The upshot of all that, is that instead of seeing a blank room with a spinner (or an error message), you can now see the basic information about the room, as well as receive incoming messages, within a few seconds of trying to join the room.
+
+## Caveats
+
+The devil, as they say, is in the detail, and unfortunately there remain a few rough edges to this work.
+
+Firstly, and most importantly: **we do not yet recommend enabling faster joins on important production systems**. We've done quite a lot of testing, but there are some fundamental changes to Synapse, and by their nature they are hard to test rigorously. We're pretty confident it won't eat your entire database, but stuck rooms, stuck notifications, and stuck clients, are all entirely within the realms of possibility. We'd recommend only enabling faster joins on your server if you are comfortable looking at the logs, and using the Synapse Admin API or flushing client caches to clean things up if they go wrong.
+
+Secondly: there's still some [outstanding work](https://github.com/matrix-org/synapse/issues/12994) required to support faster joins on Synapse deployments using worker processes. At present, if you try to enable them on a worker-based deployment, Synapse will refuse to start.
+
+Next: the faster-joins process requires some support from the resident server, and, since [those changes](https://github.com/matrix-org/matrix-spec-proposals/pull/3706) are not yet in the stable matrix spec, it must be explicitly enabled. We have enabled this support on the `matrix.org` homeserver, but it generally won't work elsewhere. In other words: for now, you'll only see the benefit when you join a room via a `#<name>:matrix.org` alias (or via the `matrix.org` room directory).
```

Comment:
Sanity check: has it been enabled on matrix.org? I guess it must have, for your test script to work?

Reply:
my... test script?

Yes, matrix.org has MSC3706 support enabled.

Reply:
I'm thinking of some sort of cronjob that perioidically joins a big room and tracks how long it takes---but maybe that was prior to the partial join stuff. It's moot anyway.

Reply:
Oh right yes. I disabled that because it wasn't producing useful information and was always breaking in one way or another.